### PR TITLE
Document tracking

### LINKED ISSUE #40: Implement document tracking (via copy pasta)

### PR DESCRIPTION
**Pull request by benocarlson**
Hey folks! Here's the document tracker and a test case, nearly line for line the same as the Groovy FileContentsTracker and its test.

### CODE REVIEW

**Review by NovaMachina — APPROVED** [review]

File: .gitignore [hunk 1]
```diff
@@ -1,2 +1,4 @@
 .DS_Store
-.idea/
\ No newline at end of file
+.idea/
+
+node_modules/
```
Comment (josiahsrc) [inline .gitignore:4]: Do we need to ignore node_modules here? I think this is already ignored in the client_vscode files

File: server/src/main/java/server/BazelServices.java [hunk 2]
```diff
@@ -12,31 +12,38 @@
 import org.eclipse.lsp4j.services.LanguageClientAware;
 import org.eclipse.lsp4j.services.TextDocumentService;
 import org.eclipse.lsp4j.services.WorkspaceService;
+import server.util.DocumentTracker;
```
Comment (josiahsrc) [inline server/src/main/java/server/BazelServices.java:16]: Do we want to separate our import statements distinguish third part/local packages?

```
import third.party.package1;
import third.party.package2;
import third.party.package3;

import local.package1;
import local.package2;
import local.package3;
```

Comment (josiahsrc) [reply server/src/main/java/server/BazelServices.java:16]: Maybe we can add this as part of the java styling package thing

File: server/src/main/java/server/util/DocumentTracker.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+                }
+                return builder.toString();
+            } catch (IOException e) {
+                return null;
+            } finally {
+                if (reader != null) {
+                    try {
+                        reader.close();
+                    } catch (IOException e) {
+                        // FIXME swallowed Exception
```
Comment (josiahsrc) [inline server/src/main/java/server/util/DocumentTracker.java]: Maybe we can use the Logger class here to log an error

File: server/src/main/java/server/util/Positions.java [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                        }
+                    }
+                }
+            } catch (IOException e) {
+                return -1;
+            }
+            try {
+                reader.close();
+            } catch (IOException e) {
+                // FIXME swallowed exception
```
Comment (josiahsrc) [inline server/src/main/java/server/util/Positions.java]: Same thing with logger

File: server/src/main/java/server/util/Positions.java [hunk 5]
```diff
@@ -0,0 +1,54 @@
+package server.util;
+
+import org.eclipse.lsp4j.Position;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.StringReader;
+import java.util.Comparator;
+
+public class Positions {
```
Comment (josiahsrc) [inline server/src/main/java/server/util/Positions.java:12]: Test case for positions class?

File: server/src/test/java/server/util/DocumentTrackerTest.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+        TextDocumentContentChangeEvent changeEvent = new TextDocumentContentChangeEvent();
+        changeEvent.setText("random nonsense");
+        changeParams.setContentChanges(Collections.singletonList(changeEvent));
+        tracker.didChange(changeParams);
+        Assertions.assertEquals("random nonsense", tracker.getContents(URI.create("test.txt")));
+    }
+
+    @Test
+    void testDidChangeWithRangeMultiline() {
+        DidOpenTextDocumentParams openParams = new DidOpenTextDocumentParams();
```
Comment (josiahsrc) [inline server/src/test/java/server/util/DocumentTrackerTest.java:62]: This code feels a little bunched up, could we use new-lines to separate related pieces in method bodies where possible? E.g.

```
@Test
    void testDidChangeWithRangeMultiline() {
        DidOpenTextDocumentParams openParams = new DidOpenTextDocumentParams();
        openParams.setTextDocument(new TextDocumentItem("multiline.txt", "plaintext", 1, "multiple\nlines"));
        tracker.didOpen(openParams);

        DidChangeTextDocumentParams changeParams = new DidChangeTextDocumentParams();
        changeParams.setTextDocument(new VersionedTextDocumentIdentifier("multiline.txt", 2));

        TextDocumentContentChangeEvent changeEvent = new TextDocumentContentChangeEvent();
        changeEvent.setText("stening\near");
        changeEvent.setRange(new Range(new Position(1, 2), new Position(1, 4)));
        changeParams.setContentChanges(Collections.singletonList(changeEvent));

        tracker.didChange(changeParams);
        Assertions.assertEquals("multiple\nlistening\nears", tracker.getContents(URI.create("multiline.txt")));
    }
```

If that kinda makes sense

**Review by josiahsrc — CHANGES_REQUESTED** [review]
Looks good overall! I left some general comments, feel free to fight me on any of them. Also if we're doing documentation, could we document public members on these changes?

File: .gitignore (hunk 1 above)
Comment (benocarlson) [reply .gitignore:4]: it tries to track it when I don't include it

File: server/src/main/java/server/util/Positions.java (hunk 5 above)
Comment (benocarlson) [reply server/src/main/java/server/util/Positions.java:12]: The groovy one doesn't seem to have tests. I guess that's no excuse, though

**Review by josiahsrc — APPROVED** [review]
lgtm

**Pull request merged**